pkg/tfbridge/check_failures.go

```diff
@@ -0,0 +1,350 @@
+	}
+	return plugin.CheckFailure{
+		Reason: reason,
+	}
+}
+
+func formatDefaultProviderCheckFailure(
+	reasonType CheckFailureReason,
+	reason string,
+	pp *CheckFailurePath,
+	configPrefix string,
+	schemaMap shim.SchemaMap,
+	schemaInfos map[string]*SchemaInfo,
+) plugin.CheckFailure {
+	if pp != nil {
+		getExpr := "pulumi config get " + pulumiConfigExpr(configPrefix, *pp)
+		reason = fmt.Sprintf("%s. Check `%s`.", reason, getExpr)
+	}
+	if pp != nil && reasonType == InvalidKey {
+		if sugg := keySuggestions(*pp, schemaMap, schemaInfos); len(sugg) > 0 {
+			quoted := []string{}
+			for _, s := range sugg {
+				quoted = append(quoted, fmt.Sprintf("`%s:%s`", configPrefix, string(s)))
+			}
+			reason = fmt.Sprintf("%s Did you mean %s?", reason, strings.Join(quoted, " or "))
+		}
+	}
+	return plugin.CheckFailure{
+		Reason: reason,
+		// Do not populate Property here intentionally as currently CLI would display that with a default
+		// provider URN as if it is a resource, which is confusing. Instead Reason contains instructions on how
+		// to set the value via a `pulumi config` command.
+	}
+}
+
+func keySuggestions(
+	pp CheckFailurePath,
+	schemaMap shim.SchemaMap,
+	schemaInfos map[string]*SchemaInfo,
+) []resource.PropertyKey {
+	// Only supported for top-level keys for now.
+	if pp.len() != 1 {
+		return nil
+	}
+	k, ok := pp.topLevelPropertyKey(schemaMap, schemaInfos)
+	if !ok {
+		return nil
+	}
+	var allKeys []resource.PropertyKey
+	schemaMap.Range(func(key string, value shim.Schema) bool {
+		translatedKey := TerraformToPulumiNameV2(key, schemaMap, schemaInfos)
+		allKeys = append(allKeys, resource.PropertyKey(translatedKey))
+		return true
+	})
+	similar := []resource.PropertyKey{}
+	for _, c := range allKeys {
+		if levenshtein.Distance(string(k), string(c), levenshtein.NewParams()) <= 2 {
+			similar = append(similar, c)
+		}
+	}
+	return similar
+}
+
+func missingDefaultProviderKey(
+	configPrefix string,
+	pp *CheckFailurePath,
+	schemaMap shim.SchemaMap,
+) plugin.CheckFailure {
+	reason := "Provider is missing a required configuration key"
+	if pp != nil {
+		configSetCommand := "pulumi config set " + pulumiConfigExpr(configPrefix, *pp)
+		reason = fmt.Sprintf("%s, try `%s`", reason, configSetCommand)
+		desc := lookupDescription(*pp, schemaMap)
+		if desc != "" {
+			reason += ": " + desc
+		}
+	}
+	return plugin.CheckFailure{
+		Reason: reason,
+		// Do not populate Property here intentionally as currently CLI would display that with a default
+		// provider URN as if it is a resource, which is confusing. Instead Reason contains instructions on how
+		// to set the value via a `pulumi config` command.
+	}
+}
+
+func pulumiConfigExpr(configPrefix string, pp CheckFailurePath) (expr string) {
+	if pp.len() > 1 {
+		expr = fmt.Sprintf("--path %s:%s", configPrefix, pp.valuePath)
+	} else {
+		expr = fmt.Sprintf("%s:%s", configPrefix, pp.valuePath)
+	}
+	return
+}
```

Comment:
Why not simplify out the `expr` and floating return?

```suggestion
func pulumiConfigExpr(configPrefix string, pp CheckFailurePath) string {
	if pp.len() > 1 {
		return fmt.Sprintf("--path %s:%s", configPrefix, pp.valuePath)
	}
	return fmt.Sprintf("%s:%s", configPrefix, pp.valuePath)
}
```